diff --git a/files/opencs.desktop b/files/opencs.desktop
index f6aad5b..80afa26 100644
--- a/files/opencs.desktop
+++ b/files/opencs.desktop
@@ -3,6 +3,7 @@ Type=Application
 Name=OpenMW Content Editor
 GenericName=Content Editor
 Comment=A replacement for the Morrowind Construction Set.
+Keywords=Morrowind;Construction Set;Creation Kit Editor;Set;Kit
 TryExec=opencs
 Exec=opencs
 Icon=opencs
diff --git a/files/openmw.desktop b/files/openmw.desktop
index 118cd3b..3e26018 100644
--- a/files/openmw.desktop
+++ b/files/openmw.desktop
@@ -3,6 +3,7 @@ Type=Application
 Name=OpenMW Launcher
 GenericName=Role Playing Game
 Comment=An engine replacement for The Elder Scrolls III: Morrowind
+Keywords=Morrowind;Reimplementation Mods;esm;bsa
 TryExec=omwlauncher
 Exec=omwlauncher
 Icon=openmw
