Description: use system libtinyxml
Author: Scott Howard <showard@debian.org>
Index: openmw/extern/oics/CMakeLists.txt
===================================================================
--- openmw.orig/extern/oics/CMakeLists.txt	2013-10-21 14:40:33.232041445 -0400
+++ openmw/extern/oics/CMakeLists.txt	2013-10-21 14:40:33.224041445 -0400
@@ -9,12 +9,10 @@
 	ICSInputControlSystem_keyboard.cpp
 	ICSInputControlSystem_mouse.cpp
 	ICSInputControlSystem_joystick.cpp
-	tinyxml.cpp
-	tinyxmlparser.cpp
-	tinyxmlerror.cpp
-    tinystr.cpp
 )
 
 add_library(${OICS_LIBRARY} STATIC ${OICS_SOURCE_FILES})
 
 link_directories(${CMAKE_CURRENT_BINARY_DIR})
+
+target_link_libraries(${OICS_LIBRARY} tinyxml)
Index: openmw/extern/oics/ICSPrerequisites.h
===================================================================
--- openmw.orig/extern/oics/ICSPrerequisites.h	2013-10-17 13:07:53.795213990 -0400
+++ openmw/extern/oics/ICSPrerequisites.h	2013-10-21 14:41:33.264042872 -0400
@@ -37,7 +37,7 @@
 #include <list>
 #include <limits>
 
-#include "tinyxml.h"
+#include <tinyxml.h>
 
 #include "SDL_keyboard.h"
 #include "SDL_mouse.h"
