Description: use system fonts instead of installing new ones
Author: Scott Howard <showard@debian.org>
Index: openmw/files/mygui/CMakeLists.txt
===================================================================
--- openmw.orig/files/mygui/CMakeLists.txt	2013-10-17 13:07:53.803213991 -0400
+++ openmw/files/mygui/CMakeLists.txt	2013-10-17 14:51:09.667361244 -0400
@@ -8,7 +8,6 @@
     black.png
     core.skin
     core.xml
-    EBGaramond-Regular.ttf
     Obliviontt.zip
     openmw_alchemy_window.layout
     openmw_book.layout
@@ -82,7 +81,6 @@
     openmw_repair.layout
     openmw_companion_window.layout
     smallbars.png
-    DejaVuLGCSansMono.ttf
     markers.png
     ../launcher/images/openmw.png
 )
Index: openmw/files/mygui/openmw_font.xml
===================================================================
--- openmw.orig/files/mygui/openmw_font.xml	2013-10-17 13:07:53.811213991 -0400
+++ openmw/files/mygui/openmw_font.xml	2013-10-17 14:52:01.071362466 -0400
@@ -24,7 +24,7 @@
     </Resource>
 
     <Resource type="ResourceTrueTypeFont" name="MonoFont">
-        <Property key="Source" value="DejaVuLGCSansMono.ttf"/>
+        <Property key="Source" value="DejaVuSansMono.ttf"/>
         <Property key="Size" value="18"/>
         <Property key="Resolution" value="50"/>
         <Property key="Antialias" value="false"/>
